src/mem/freelist.h

```diff
@@ -392,53 +411,56 @@ namespace snmalloc
      */
     FreeListIter terminate(LocalEntropy& entropy, bool preserve_queue = true)
     {
-      SNMALLOC_ASSERT(end[1] != &head[0]);
-      SNMALLOC_ASSERT(end[0] != &head[1]);
-
-      // If second list is empty, then append is trivial.
-      if (end[1] == &head[1])
+      if constexpr (RANDOM)
       {
-        end[0]->store(nullptr, get_prev(0), entropy);
-        return {head[0].read(HEAD_KEY, entropy)};
-      }
-
-      end[1]->store(nullptr, get_prev(1), entropy);
+        SNMALLOC_ASSERT(end[1] != &head[0]);
+        SNMALLOC_ASSERT(end[0] != &head[1]);
 
-      // Append 1 to 0
-      auto mid = head[1].read(HEAD_KEY, entropy);
-      end[0]->store(mid, get_prev(0), entropy);
-      // Re-code first link in second list (if there is one).
-      // The first link in the second list will be encoded with initial_key,
-      // But that needs to be changed to the curr of the first list.
-      if (mid != nullptr)
-      {
-        auto mid_next = mid->read_next(initial_key(mid) & 0xffff, entropy);
-        mid->next_object.store(mid_next, get_curr(0), entropy);
-      }
-
-      auto h = head[0].read(HEAD_KEY, entropy);
-
-      // If we need to continue adding to the builder
-      // Set up the second list as empty,
-      // and extend the first list to cover all of the second.
-      if (preserve_queue && h != nullptr)
-      {
-#ifdef CHECK_CLIENT
-        prev[0] = prev[1];
-        curr[0] = curr[1];
-#endif
-        end[0] = end[1];
-#ifdef CHECK_CLIENT
-        prev[1] = HEAD_KEY;
-        curr[1] = initial_key(h) & 0xffff;
-#endif
-        end[1] = &(head[1]);
+        // If second list is non-empty, perform append.
+        if (end[1] != &head[1])
+        {
+          end[1]->store(nullptr, get_prev(1), entropy);
+
+          // Append 1 to 0
```

Comment:
Possibly worth refactoring out to a method for appending these encoded lists?  If you think it won't get used anywhere else I'm happy to leave it in place, tho'.

Reply:
Generally I'm in favour of doing things like this, but I am not sure this will add much value as almost all the code is about appending, or depends on precisely what the append does.